chiselFrontend/src/main/scala/chisel3/core/StrongEnum.scala

```diff
@@ -0,0 +1,290 @@
+        case e :: Nil => first_enum
+        case e :: e_next :: es => Mux(this === e, e_next, muxBuilder(e_next :: es, first_enum))
+      }
+
+      muxBuilder(factory.all.toList, factory.all.head).asInstanceOf[this.type]
+    }
+  }
+
+  private[core] def bindToLiteral(num: BigInt, w: Width): Unit = {
+    val lit = ULit(num, w)
+    lit.bindLitArg(this)
+  }
+
+  override def bind(target: Binding, parentDirection: SpecifiedDirection): Unit = {
+    super.bind(target, parentDirection)
+
+    // If we try to annotate something that is bound to a literal, we get a FIRRTL annotation exception.
+    // To workaround that, we only annotate enums that are not bound to literals.
+    if (selfAnnotating && !litOption.isDefined) {
+      annotateEnum()
+    }
+  }
+
+  private def annotateEnum(): Unit = {
+    annotate(EnumComponentChiselAnnotation(this, enumTypeName))
+
+    if (!Builder.annotations.contains(factory.globalAnnotation)) {
+      annotate(factory.globalAnnotation)
+    }
+  }
+
+  protected def enumTypeName: String = factory.enumTypeName
+
+  def toPrintable: Printable = FullName(this) // TODO: Find a better pretty printer
+}
+
+
+abstract class EnumFactory {
+  class Type extends EnumType(this)
+  object Type {
+    def apply(): Type = EnumFactory.this.apply()
+  }
+
+  var id: BigInt = 0
+  var width: Width = 0.W
+
+  val enum_names = mutable.ArrayBuffer.empty[String]
+  val enum_values = mutable.ArrayBuffer.empty[BigInt]
+  val enum_instances = mutable.ArrayBuffer.empty[Type]
+
+  private[core] def globalAnnotation: EnumDefChiselAnnotation =
+    EnumDefChiselAnnotation(enumTypeName, (enum_names, enum_values).zipped.toMap)
+
+  private[core] val enumTypeName = getClass.getName.init
+
+  def getWidth: Int = width.get
+
+  def all: Seq[Type] = enum_instances.toSeq
+
+  protected def Value: Type = macro EnumMacros.ValImpl
+  protected def Value(id: UInt): Type = macro EnumMacros.ValCustomImpl
+
+  protected def do_Value(names: Seq[String]): Type = {
+    val result = new Type
+    enum_names ++= names.filter(!enum_names.contains(_))
+    enum_instances.append(result)
+    enum_values.append(id)
+
+    // We have to use UnknownWidth here, because we don't actually know what the final width will be
+    result.bindToLiteral(id, UnknownWidth())
+
+    width = (1 max id.bitLength).W
+    id += 1
+
+    result
+  }
+
+  protected def do_Value(names: Seq[String], id: UInt): Type = {
+    // TODO: These throw ExceptionInInitializerError which can be confusing to the user. Get rid of the error, and just
+    // throw an exception
+    if (!id.litOption.isDefined)
+      throwException(s"$enumTypeName defined with a non-literal type")
+    if (id.litValue() < this.id)
+      throwException(s"Enums must be strictly increasing: $enumTypeName")
+
+    this.id = id.litValue()
+    do_Value(names)
+  }
+
+  def apply(): Type = new Type
+
+  def apply(n: UInt)(implicit sourceInfo: SourceInfo, connectionCompileOptions: CompileOptions): Type = {
```

Comment:
Might be worth having a discussion on apply vs. fromBits. Especially since fromBits gives you a warning anyways.

Reply:
As per our discussion, I've moved `fromBits` into `apply`. Enums will be experimental anyway, so we've got time to refine this.